web3.js/src/util/send-and-confirm-transaction.ts

```diff
@@ -34,36 +40,64 @@ export async function sendAndConfirmTransaction(
     sendOptions,
   );
 
-  const status = await connection.getSignatureStatus(signature);
+  const subscriptionCommitment =
+    options?.preflightCommitment || options?.commitment;
+  let subscriptionId: number | undefined;
+  let status: RpcResponseAndContext<SignatureResult> | null = null;
+
+  const confirmTx = new Promise((resolve, reject) => {
+    try {
+      subscriptionId = connection.onSignature(
+        signature,
+        (result: SignatureResult, context: Context) => {
+          subscriptionId = undefined;
+          status = {
+            context,
+            value: result,
+          };
+          resolve(null);
+        },
+        subscriptionCommitment,
+      );
+    } catch (err) {
+      reject(err);
+    }
+  });
 
   const checkBlockHeight = async () => {
     const blockHeight = await connection.getBlockHeight(
       options && options.commitment,
     );
 
-    if (blockHeight > transaction.lastValidBlockHeight!) {
-      throw new Error('Transaction has expired.');
-    } else {
-      if (status) {
-        return;
+    return blockHeight;
+  };
+
+  let currentBlockHeight = await checkBlockHeight();
+
+  const expireTx = async (): Promise<void> => {
+    while (currentBlockHeight <= transaction.lastValidBlockHeight!) {
+      if (transaction.lastValidBlockHeight! - currentBlockHeight > 200) {
+        await sleep(2000);
+      } else if (transaction.lastValidBlockHeight! - currentBlockHeight > 100) {
+        await sleep(1000);
+      } else {
+        await sleep(500);
       }
-      await sleep(500);
-      await checkBlockHeight();
+      currentBlockHeight = await checkBlockHeight();
     }
   };
```

Comment:
That's what we originally set out to do, but it may not have been the best approach. We made the recursive call and then exited once `getSignatureStatus` was successful, otherwise it would throw an error if the `blockHeight` exceeded `lastValidBlockHeight`. For this refactor, we decided to make it so this function was only responsible for monitoring where the `blockHeight` is in relation to `lastValidBlockHeight`, passing both of these to the `promiseExpiration` function. Looking at this now, I see I forgot to do something in the event the loop is exited, which should have been something about letting the user know the transaction expired. Taking all of your suggestions into account, I think this can be cleaned up quite a bit.

Reply:
Yeah! It's totally fine for the _last_ fetch to finish, just don't start new ones once we get a signature confirmation.